File: P12977.py
def get_sieve():
    
    res = [1] * 3000
    res[0] = 0
    res[1] = 0
    
    for i in range(4, 3000, 2):
        res[i] = 0
    for i in range(3, 3000, 2):
        if res[i]:
            for j in range(2*i, 3000, i):
                res[j] = 0
    
    return res

def solution(nums):
    
    answer = 0

    sieve = get_sieve()
    N = len(nums)
    
    for i in range(N-2):
        for j in range(i+1, N-1):
            for k in range(j+1, N):
                if sieve[nums[i]+nums[j]+nums[k]]:
                    answer += 1

    return answer

File: P12977.java
public class P12977 {
    int answer = 0;
    boolean[] sieve = new boolean[3001];

    public int solution(int[] nums) {

        this.sieve[0] = false;
        this.sieve[1] = false;
        this.sieve[2] = true;
        for (int i=4; i<3001; i+=2) this.sieve[i] = false;
        for (int i=3; i<3001; i+=2) this.sieve[i] = true;
        for (int i=3; i<3001; i+=2) {
            if (this.sieve[i]) {
                for (int j=2*i; j<3001; j+=i) this.sieve[j] = false;
            }
        }

        dfs(nums, 0, 0, 0);

        return this.answer;
    }

    public void dfs(int[] nums, int idx, int depth, int total) {
        
        if (depth == 3) {
            if (this.sieve[total]) this.answer += 1;
            return;
        }

        if (depth > 3 || idx == nums.length) return;

        dfs(nums, idx+1, depth+1, total+nums[idx]);
        dfs(nums, idx+1, depth, total);
    }
}
